Add facility for hhvm macros to devel package to ease packaging of extensions

samveen opened the pull request:
Also fixed minor naming issue with logrotate source in %install section of hhvm.spec .

This is paticularly important as the path used in the hhvm spec is different from the path used in extension specs:

From [hhvm-repo/hhvm/hhvm.spec](https://github.com/no1youknowz/hhvm-repo/blob/master/hhvm/hhvm.spec#L30), line 30
```%define hhvm_extensiondir %{_libdir}/hhvm/extensions/%{hhvm_api_version}```

Compare this to [hhvm-repo/hhvm-aerospike/hhvm-aerospike.spec](https://github.com/no1youknowz/hhvm-repo/blob/master/hhvm-aerospike/hhvm-aerospike.spec#L34), line 34
```%{__mkdir} -p %{buildroot}/usr/lib64/hhvm/extensions```

AFAICS that's a maintenance nightmare in the making.

jrobeson commented on hhvm/macros.hhvm:
```diff
@@ -0,0 +1,16 @@
+# Binaries
+%__hphpize HPHPIZE
+
+# Convenience macros - Test this - Copied from macros.cmake
```
It says test this. Have you tested it?

jrobeson commented on hhvm/macros.hhvm:
```diff
@@ -0,0 +1,16 @@
+%__hphpize HPHPIZE
+
+# Convenience macros - Test this - Copied from macros.cmake
+%__cmake_hhvm cmake -DCMAKE_INSTALL_LIBDIR:PATH=%{_hhvm_extensiondir}
```
do we really wanna call it __cmake anything? maybe just %__makeext or something to that effect. I forgot what Fedora uses in the php sec, but it'd probably be best to follow that.

jrobeson commented on hhvm/macros.hhvm:
```diff
@@ -0,0 +1,16 @@
+# There are to be used for packaging of extensions that use hhvm
+
+# Versioning info
+%_hhvm_api_version HHVM_API_VERSION
```
We definitely need these version macros, but i hesitated to implement them until we actually verify that the numbers are correct when we update hhvm like http://pkgs.fedoraproject.org/cgit/php.git/tree/php.spec?h=f22#n760

jrobeson replied:
do we really need the _ prefix?

jrobeson commented on hhvm/hhvm.spec:
```diff
@@ -277,27 +278,22 @@ make install DESTDIR=%{buildroot}
 execstack -c %{buildroot}%{_bindir}/hhvm
 
-mkdir -p %{buildroot}%{hhvm_extensiondir}
+%{__mkdir_p} %{buildroot}%{hhvm_extensiondir}
```
do using these provide any benefit? It's Fedora policy (and thus RHEL) not to use use these macros by default.

jrobeson commented on hhvm/hhvm.spec:
```diff
@@ -307,14 +303,28 @@ install -p -D -m 644 %{SOURCE5} %{buildroot}%{_sysconfdir}/nginx/default.d/hhvm.
+%{__sed} -i "s#HHVM_PHP_API_VERSION#%{php_api_version}#" %{buildroot}%{_sysconfdir}/rpm/macros.hhvm
+%{__sed} -i "s#HHVM_EXTENSION_DIR#%{hhvm_extensiondir}#" %{buildroot}%{_sysconfdir}/rpm/macros.hhvm
+%{__sed} -i "s#HPHPIZE#%{_bindir}/hphpize#" %{buildroot}%{_sysconfdir}/rpm/macros.hhvm
+
```
see my comment below: https://github.com/no1youknowz/hhvm-repo/pull/3/files#r31018708

samveen commented:
@jrobeson I've modified the PR to follow the [Fedora Packaging guidelines](https://fedoraproject.org/wiki/Packaging:Guidelines) more closely. Thank you for pointing that out. 

I have also removed the cmake macro ( I still need to do more tests on it).

As for the values of the macros, they are inherited from the values in `hhvm.spec` (replaced in `macros.hhvm`  using `sed` at package build time).

jrobeson commented:
The macros for versions need a sanity check to make sure we bump the versions when hhvm does. That's why i linked to that specific line(s) for the php.spec.

samveen commented:
HHVM PHP Version: I searched through the source for `PHP_VERSION` and found definitions at the following places:
1. hphp/system/idl/constants.idl.json
2. hphp/runtime/ext_zend_compat/php-src/main/php_version.h

I'm unsure as to which is the primary version.

HHVM API version:  `runtime/ext/extension.h`

HHVM PHP API version: 
runtime/ext_zend_compat/php-src/main/php.h

Thinking long term, most of the hhvm provided macros probably need to match the [macros provided by the php package](http://pkgs.fedoraproject.org/cgit/php.git/tree/macros.php).

How do you envision the sanity check implemention?

jrobeson commented:
the idl machinery is going away, so as long as php_version.h isn't behind the one in constants.idl.json, then we should use that.

I envisioned the sanity check in a similar way as the php.spec for Fedora does it. Check these lines: http://pkgs.fedoraproject.org/cgit/php.git/tree/php.spec?h=f22#n760

samveen commented:
There is a mismatch:
```
[samveen@development hhvm]$ f=hphp/system/idl/constants.idl.json; grep -n PHP_VERSION $f |head -n1 |cut -f1 -d: |xargs -I{} bash -c "tail -n +{} $f |head -n2 |tail -n1|cut -f4 -d\\\""
5.6.99-hhvm
```
versus
```
[samveen@development hhvm]$ f=hphp/runtime/ext_zend_compat/php-src/main/php_version.h ;grep -n '\<PHP_VERSION\>' $f|cut -f2 -d\"
5.5.5-dev
```

jrobeson commented:
Could be. Or we just don't know the right one and i was just making an assumption. I guess we'll have to ask the hhvm folks on irc.

jrobeson commented:
I think i figured it out. For now we want to use the one in constants.idl.json. In the current hhvm master, it's been moved to its final location in hphp/hack/hhi/constants.hhi , so we'll adjust to that for 2.8.

samveen commented:
@jrobeson Added the checks . I also added a version check (this isn't for the macro file, but for the hhvm version itself)

jrobeson commented on hhvm/hhvm.spec:
```diff
@@ -307,6 +336,20 @@ install -p -D -m 644 %{SOURCE5} %{buildroot}%{_sysconfdir}/nginx/default.d/hhvm.
+sed -i "s#HHVM_PHP_API_VERSION#%{php_api_version}#" %{buildroot}%{_rpmconfigdir}/macros.d/macros.hhvm
+sed -i "s#HHVM_EXTENSION_DIR#%{hhvm_extensiondir}#" %{buildroot}%{_rpmconfigdir}/macros.d/macros.hhvm
+sed -i "s#HPHPIZE#%{_bindir}/hphpize#" %{buildroot}%{_rpmconfigdir}/macros.d/macros.hhvm
+
```
I don't think we should be touching files in their target locations here.

Can you do it like it's done here? http://pkgs.fedoraproject.org/cgit/php.git/tree/php.spec?h=f22#n1320

samveen replied:
@jrobeson  I'll rename `macros.hhvm` to `macros.hhvm.in` (following the autoconf convention), if you don't have any objections.

I'll implement the rest of the changes from there.

jrobeson replied:
sure, why not.

samveen replied:
Done

samveen commented:
I tested the macros with the hhvm-ext-uuid spec and it works as expected. Once you merge this commit, I'll work on changing each extension's spec to work with the macros (individual pull requests per extension).

jrobeson commented:
I really appreciate your work on this. Thanks very much!

jrobeson commented:
Oh.. i should have had you squash it first.

jrobeson commented:
I have one other request. Can you prefix future PRs/commits with the module name? like [package-name] ?

samveen commented:
Will do. (And I'll read up on commit squashing as well :) )
